File: src/core/differ.ts
import type { ActiveMcp, ScannedMcp, DiffEntry } from "../types.js";

export function computeDiff(
  desired: ActiveMcp[],
  current: ScannedMcp[]
): DiffEntry[] {
  const diffs: DiffEntry[] = [];

  const desiredSet = new Set<string>();
  for (const mcp of desired) {
    const scopes: Array<"global" | "project"> =
      mcp.scope === "both" ? ["global", "project"] : [mcp.scope];

    for (const scope of scopes) {
      for (const agent of mcp.agents) {
        const key = `${mcp.name}|${agent}|${scope}`;
        desiredSet.add(key);

        const exists = current.some(
          (c) => c.name === mcp.name && c.source.agent === agent && c.source.scope === scope
        );

        if (!exists) {
          diffs.push({ action: "add", mcpName: mcp.name, agent, scope });
        }
      }
    }
  }

  for (const c of current) {
    const key = `${c.name}|${c.source.agent}|${c.source.scope}`;
    if (!desiredSet.has(key)) {
      diffs.push({
        action: "remove",
        mcpName: c.name,
        agent: c.source.agent,
        scope: c.source.scope,
      });
    }
  }

  return diffs;
}

File: desktop/src-tauri/src/core/differ.rs
#[derive(Debug, Clone, PartialEq)]
pub enum DiffAction { Add, Remove }

#[derive(Debug, Clone, PartialEq)]
pub struct DiffEntry {
    pub action: DiffAction,
    pub mcp_name: String,
    pub agent: String,
    pub scope: String,
}

#[derive(Debug, Clone)]
pub struct DesiredMcp {
    pub name: String,
    pub agents: Vec<String>,
    pub scopes: Vec<String>, // expanded global/project
}

#[derive(Debug, Clone)]
pub struct CurrentMcp {
    pub name: String,
    pub agent: String,
    pub scope: String,
}

pub fn compute_diff(desired: &[DesiredMcp], current: &[CurrentMcp]) -> Vec<DiffEntry> {
    let mut diffs = Vec::new();
    let mut desired_set = std::collections::HashSet::new();
    for d in desired {
        for scope in &d.scopes {
            for agent in &d.agents {
                let key = format!("{}|{}|{}", d.name, agent, scope);
                desired_set.insert(key.clone());
                let exists = current.iter().any(|c|
                    c.name == d.name && &c.agent == agent && &c.scope == scope);
                if !exists {
                    diffs.push(DiffEntry { action: DiffAction::Add,
                        mcp_name: d.name.clone(), agent: agent.clone(), scope: scope.clone() });
                }
            }
        }
    }
    for c in current {
        let key = format!("{}|{}|{}", c.name, c.agent, c.scope);
        if !desired_set.contains(&key) {
            diffs.push(DiffEntry { action: DiffAction::Remove,
                mcp_name: c.name.clone(), agent: c.agent.clone(), scope: c.scope.clone() });
        }
    }
    diffs
}

#[cfg(test)]
mod tests {
    use super::*;
    #[test]
    fn adds_missing_and_removes_extra() {
        let desired = vec![DesiredMcp {
            name: "git".into(), agents: vec!["claude-code".into()],
            scopes: vec!["global".into()] }];
        let current = vec![CurrentMcp {
            name: "old".into(), agent: "claude-code".into(), scope: "global".into() }];
        let diffs = compute_diff(&desired, &current);
        assert!(diffs.iter().any(|d| d.action == DiffAction::Add && d.mcp_name == "git"));
        assert!(diffs.iter().any(|d| d.action == DiffAction::Remove && d.mcp_name == "old"));
    }
}
